### PR TITLE
Added a couple of funtions to handle the ability to use libsqlcipher

### PR DESCRIPTION
**Pull request by kodeschreiber**
- Added the "sqlite3_key()" function
- Added the "sqlite3_rekey()" function
- Added a feature gate for encryption

### CODE REVIEW

**Comment by IvanUkhov** [comment]
I have been trying to recall how I got those functions into the sys crate. Looks like there were present in the past but are long gone in the current version of SQLite. So I plan to do a sync at some point, and those functions will disappear. SQLCipher is a standalone project.

**Comment by kodeschreiber** [comment]
Yeah, my understanding was that the encryption that was/is baked into SQLite is a paid feature. The SQLCipher project is a community version of that feature. Given this, I totally understand if this is out-of-scope, since it's not technically available in the stock public SQLite library. I can always make a separate supporting library instead. Feel free to close this PR if this is the case.

**Comment by IvanUkhov** [comment]
Hm, I was not aware of this. I was more thinking that they extracted that part into a standalone project, not that they close-sourced it.

Do you know if there is a one-to-one correspondence between the two? I am thinking about what to do with the function definitions. Ideally, we only want what is in that closed part of the official SQLite.

**Comment by kodeschreiber** [comment]
There is some overlap, but not entirely:
[SQLite3 SEE - closed-soruce](https://www.sqlite.org/see/doc/trunk/www/readme.wiki)
[SQLCipher](https://github.com/sqlcipher/sqlcipher)

They both have the "sqlite3_key" and "sqlite3_rekey" functions, but the SEE library has "sqlite3_key_v2" and "sqlite3_rekey_v2". SEE also has "sqlite3_activate_see". So while they have some functions in common, the official SEE version has more functions than SQLCipher. I guess the overlap is "safe" to implement (assuming the function signatures are the same). Might still be a good idea to use a feature gate.

On a similar note, the SQLite maintainers also have a paid option for their compression extension of SQLite: (CEROD)[https://sqlite.org/com/cerod.html]

**Comment by IvanUkhov** [comment]
I think we can find some middle ground here. I wonder if one can get ahold of the header file with the functions added by that extension, to have an official reference. On the page, they list only two. One can of course assume that there are only four, and that what was in the repository back then holds even today.

https://github.com/sqlite/sqlite/blob/version-3.16.2/src/sqlite.h.in#L5086-L5129

**Comment by IvanUkhov** [comment]
Updated `sqlite3-src` and `sqlite3-sys` to bring it up to speed and also accommodate the encryption. See the features:

https://github.com/stainless-steel/sqlite3-sys/tree/main

File: src/connection.rs [hunk 1]
```diff
@@ -225,6 +225,44 @@ impl Connection {
         Ok(())
     }
 
+
+    #[cfg(feature = "encryption")]
```
Comment (IvanUkhov) [inline src/connection.rs:229]: It requires a docstring:

```
/// Set the encryption key.
```

File: src/connection.rs [hunk 2]
```diff
@@ -225,6 +225,44 @@ impl Connection {
...
+    #[inline]
+    pub fn encryption_key<T: AsRef<str>>(&self, data: T) -> Result<()> {
+        unsafe {
+            ok!(
+                self.raw.0,
+                ffi::sqlite3_key(
```
Comment (IvanUkhov) [inline src/connection.rs:235]: Should we switch to version 2 perhaps? Seems to be supported by both.

https://www.zetetic.net/sqlcipher/sqlcipher-api/#sqlite3_key

File: src/connection.rs [hunk 3]
```diff
@@ -225,6 +225,44 @@ impl Connection {
...
+                    self.raw.0,
+                    str_to_cstr!(data.as_ref()).as_ptr() as *const c_void,
+                    data.len() as c_int,
+                )
+            );
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "encryption")]
```
Comment (IvanUkhov) [inline src/connection.rs:245]:
```
/// Change the encryption key.
```

File: src/connection.rs [hunk 4]
```diff
@@ -225,6 +225,44 @@ impl Connection {
...
+    #[inline]
+    pub fn encryption_rekey<T: AsRef<str>>(&self, data: T) -> Result<()> {
+        unsafe {
+            ok!(
+                self.raw.0,
+                ffi::sqlite3_key(
```
Comment (IvanUkhov) [inline src/connection.rs:251]: Type here.

Comment (IvanUkhov) [reply src/connection.rs:251]: Should we switch to version 2?

https://www.zetetic.net/sqlcipher/sqlcipher-api/#sqlite3_rekey

File: src/connection.rs [hunk 5]
```diff
@@ -225,6 +225,44 @@ impl Connection {
...
+                    str_to_cstr!(data.as_ref()).as_ptr() as *const c_void,
+                    data.len() as c_int,
+                )
+            );
+        }
+        Ok(())
+    }
+
+
+// int sqlite3_key(sqlite3 *db, const void *pKey, int nKey);
```
Comment (IvanUkhov) [inline src/connection.rs:262]: Please remove comments and excessive empty lines.

File: src/connection.rs [hunk 6]
```diff
@@ -225,6 +225,44 @@ impl Connection {
         Ok(())
     }
 
+
+    #[cfg(feature = "encryption")]
+    #[inline]
+    pub fn encryption_key<T: AsRef<str>>(&self, data: T) -> Result<()> {
```
Comment (IvanUkhov) [inline src/connection.rs:231]: Let us rename to `set_encryption_key`.

File: src/connection.rs [hunk 7]
```diff
@@ -225,6 +225,44 @@ impl Connection {
...
+                    data.len() as c_int,
+                )
+            );
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "encryption")]
+    #[inline]
+    pub fn encryption_rekey<T: AsRef<str>>(&self, data: T) -> Result<()> {
```
Comment (IvanUkhov) [inline src/connection.rs:247]: Let us rename to `change_encryption_key`.

**Comment by kodeschreiber** [comment]
That all sounds good to me. I've reworked my patch to use the v2 functions (I wasn't aware that SQLCipher had those implemented as well). Also, I must have been rather tired making this patch; just noticed some of the issues I had when reviewing your suggestions :). So, between applying your changes and catching up with your latest commits, I'll just open a new PR with the updates. My last branch is a bit disorganized from experimenting.

**Pull request closed without merging**